### PR TITLE
feat(dio): Allow ResponseDecoder and RequestEncoder to be async

### PR DESCRIPTION
**Pull request by Reprevise**
### New Pull Request Checklist

- [x] I have read the [Documentation](https://pub.dev/documentation/dio/latest/)
- [x] I have searched for a similar pull request in the [project](https://github.com/cfug/dio/pulls) and found none
- [x] I have updated this branch with the latest `main` branch to avoid conflicts (via merge from master or rebase)
- [x] I have added the required tests to prove the fix/feature I'm adding
- [ ] I have updated the documentation (if necessary)
- [x] I have run the tests without failures
- [x] I have updated the `CHANGELOG.md` in the corresponding package

### Additional context and info (if any)

Gives more flexibility to the developer if they want to do some async work in the response decoder. I don't think tests are needed for this. I don't _think_ this is a breaking change as the return type of `ResponseDecoder` is now `FutureOr<String?>`.

### CODE REVIEW

File: dio/lib/src/transformers/sync_transformer.dart [hunk 1]
```diff
@@ -125,7 +125,7 @@ class SyncTransformer extends Transformer {
     );
     final String? response;
     if (options.responseDecoder != null) {
-      response = options.responseDecoder!(
+      response = await options.responseDecoder!(
```
Comment (ueman) [inline dio/lib/src/transformers/sync_transformer.dart:128]: This should not await the return value if it's not a future. The code should then roughly look like the following code:
```dart
      final responseValue = await options.responseDecoder!(...);
      if (responseValue is Future) {
        response = await responseValue;
      } else {
        response = responseValue;
      }
```

**Review by ueman — DISMISSED** [review]
I'm not sure whether this should be considered a breaking change. I would love to know what the others think.

However, this should be tested. Please add a test where the `ResponseDecoder` returns a future and one where it doesn't return a future.

**Review by kuhnroyal — APPROVED** [review]
I think this is fine. Should not break anything.

**Comment by AlexV525** [comment]
Could we address `ResponseEncoder` too? If so, https://github.com/cfug/dio/issues/1709 should be resolved.

**Comment by AlexV525** [comment]
Sorry, I mean the `RequestEncoder`. :)

**Comment by AlexV525** [comment]
> EDIT 2: I can allow RequestEncoder to return a Future in this PR, or I can open another one. Up to you guys.

@Reprevise Please continue with this PR if possible, thanks.

**Comment by Reprevise** [comment]
@AlexV525 `RequestEncoder` has been updated to return a FutureOr<List<int>>

**Comment by AlexV525** [comment]
You seem to be using a conflicted main branch, which merges your changes unexpectedly.

**Comment by Reprevise** [comment]
Should be fixed now.

**Review by AlexV525 — APPROVED** [review]
LGTM. You'll need to update the CHANGELOG before it gets merged.

**Comment by Reprevise** [comment]
Changelog has been updated

**Pull request merged**